src/main/java/org/ambraproject/wombat/config/theme/ThemeBuilder.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2017 Public Library of Science
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package org.ambraproject.wombat.config.theme;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * An object that contains all the data necessary to construct a {@link Theme} object.
+ * <p>
+ * The purpose of this intermediate step is to allow a {@code Theme}'s parents to be constructed first. Because a {@code
+ * Theme} has immutable pointers to its parents, they must all be constructed in one pass while building a {@link
+ * ThemeTree}. Code that parses theme specifications can represent it as {@code ThemeBuilder} objects because a {@code
+ * ThemeBuilder} only needs to refer to parents by their key string, not as objects.
+ * <p>
+ * A {@link ConstructorFunction} object encapsulates whatever additional values are necessary to construct a theme of
+ * the particular subtype. For example, for a {@link FileTheme}, the {@code ConstructorFunction} would hold on to a
+ * pointer to a {@link java.io.File} object and pass it to the {@code FileTheme} constructor.
```

Comment:
Feedback requested: does this make sense?

I experimented with other ways of doing the same thing, mainly by making `ThemeBuilder` an abstract class with one extension for `FileTheme`, etc. This way wound up being much less verbose.

All this indirection is worth it so that we can separate out the theme-building step and put it closer to where we parse the configuration. This already matters for injecting themes into unit tests (see the updated tests in this PR). It may matter even more in the future if we ever want to plug in themes from somewhere other than the server's local filesystem.